### PR TITLE
TensorRT QAT - Object Detection

### PR DESCRIPTION
**Pull request by Akash-guna**
Previous PR #9

### CODE REVIEW

File: quant/tensorrt/qat/main.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        self.qat = kwargs.get("QAT", False)
+        self.platform = kwargs.get("PLATFORM", "mmdet")
+
+    def list_subdirectories(self, directory):
+        subdirectories = [
+            d
+            for d in os.listdir(directory)
+            if os.path.isdir(os.path.join(directory, d))
+        ]
+        return subdirectories
```
Comment (shwu-nyunai) [inline quant/tensorrt/qat/main.py:64]: os.walk could be used - _[ref docs](https://docs.python.org/3/library/os.html#os.walk)_

File: quant/tensorrt/qat/main.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            self.momentum,
+            self.lr,
+            self.scheduler,
+            self.factor,
+            self.weight_decay,
+        )
+        customize_config(config, self.data_path, self.model_path, self.batch_size)
+
+        os.system(
+            f"python /mmrazor/tools/train.py current_quant_final.py --work-dir {self.job_path}"
```
Comment (shwu-nyunai) [inline quant/tensorrt/qat/main.py:103]: same as above. (absolute path)

File: quant/tensorrt/qat/main.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        if self.quantized_pth_location == None:
+            self.logger.info("Fake Quantization Unsuccessful, checkpoint not found")
+            raise Exception("Fake Quantization Unsuccessful, checkpoint not found")
+        else:
+            self.logger.info("Fake Quantization Successful")
+        # deply config
+        build_mmdeploy_config(self.imsize)
+        create_input_image(self.loaders["test"])
+        os.system(
+            f"docker exec -it Nyun_Kompress python mmdeploy/tools/deploy.py current_tensorrt_deploy_config.py current_quant_final.py {self.quantized_pth_location} demo_image.png"
```
Comment (shwu-nyunai) [inline quant/tensorrt/qat/main.py:123]: why is a docker exec command used?

File: quant/tensorrt/qat/main.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        )
+        self.logger.info("Deployment Successful")
+        shutil.move("/workspace/end2end.xml", os.path.join(self.model_path, "mds.xml"))
+        shutil.move("/workspace/end2end.bin", os.path.join(self.model_path, "mds.bin"))
```
Comment (shwu-nyunai) [inline quant/tensorrt/qat/main.py:127]: absolute paths.

File: quant/tensorrt/qat/utils.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    model_inputs=[
+        dict(
+            input_shapes=dict(
+                input=dict(
+                    min_shape=[1, 3, {insize}, {insize}],
+                    opt_shape=[1, 3, {insize}, {insize}],
+                    max_shape=[1, 3, {insize}, {insize}])))
+    ])"""
+    with open("current_tensorrt_deploy_config.py", "w") as f:
+        f.write(config)
```
Comment (shwu-nyunai) [inline quant/tensorrt/qat/utils.py:170]: same as above.

**Review by shwu-nyunai — CHANGES_REQUESTED** [review]

File: quant/tensorrt/qat/main.py (hunk 1 above)
Comment (Akash-guna) [reply quant/tensorrt/qat/main.py]: Wasnt using it anywhere removed it

File: quant/tensorrt/qat/main.py (hunk 2 above)
Comment (Akash-guna) [reply quant/tensorrt/qat/main.py]: changed (will be changed to vision/core/utils/mmrazortrain.py)

File: quant/tensorrt/qat/main.py (hunk 3 above)
Comment (Akash-guna) [reply quant/tensorrt/qat/main.py]: Changed the flow , now the user needs to run the yaml twice to get a truely quantized model

File: quant/tensorrt/qat/utils.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+                        max_shape=[1, 3, {insize}, {insize}])))
+        ]),
+    function_record_to_pop=[
+        'mmdet.models.detectors.single_stage.SingleStageDetector.forward',
+        'mmdet.models.detectors.two_stage.TwoStageDetector.forward',
+        'mmdet.models.detectors.single_stage_instance_seg.SingleStageInstanceSegmentor.forward',  # noqa: E501
+        'torch.cat'
+    ])"""
+    with open("current_base_tensorrt_deploy_config.py", "w") as f:
+        f.write(cfg)
```
[the remark this answers is not in the document]
Comment (Akash-guna) [reply quant/tensorrt/qat/utils.py]: done

File: quant/tensorrt/qat/main.py [hunk 7]
```diff
@@ -44,25 +44,17 @@ def __init__(self, model, loaders=None, **kwargs):
         self.val_interval = kwargs.get("VALIDATION_INTERVAL", 1)
         self.weight_decay = kwargs.get("WEIGHT_DECAY", 0.0005)
         self.model_path = kwargs.get("MODEL_PATH", "models")
-        self.logging_path = kwargs.get("LOGGING_PATH", "logs")
-
-        self.logging_path = logging.getLogger(__name__)
+        self.work_dir = os.getcwd()
```
Comment (shwu-nyunai) [inline quant/tensorrt/qat/main.py:47]: lets derive this at runtime.

File: quant/tensorrt/qat/main.py [hunk 8]
```diff
@@ -97,33 +90,46 @@ def compress_model(self):
...
+                    )
 
-        if self.quantized_pth_location == None:
-            self.logger.info("Fake Quantization Unsuccessful, checkpoint not found")
-            raise Exception("Fake Quantization Unsuccessful, checkpoint not found")
+            if self.quantized_pth_location == None:
+                self.logger.info("Fake Quantization Unsuccessful, checkpoint not found")
+                raise Exception("Fake Quantization Unsuccessful, checkpoint not found")
+            else:
+                self.logger.info("Fake Quantization Successful")
```
Comment (shwu-nyunai) [inline quant/tensorrt/qat/main.py:113]: no need to log if the exception is anyway going to be the last thing logged by the logger

File: quant/tensorrt/qat/main.py [hunk 9]
```diff
@@ -97,33 +90,46 @@ def compress_model(self):
...
-        shutil.move("/workspace/end2end.bin", os.path.join(self.model_path, "mds.bin"))
+            # deply config
+            self.quantized_pth_location = self.kwargs.get("FAKE_QUANTIZED_PATH", "")
+
+            if self.quantized_pth_location == None:
+                self.logger.info(f"Fake Quantized Path is None")
+                raise Exception("Fake Quantized Path is None")
+            elif not os.path.exists(self.quantized_pth_location):
+                self.logger.info(f"Fake Quantized Path is Not Present at {self.quantized_pth_location}")
+                raise Exception(f"Fake Quantized Path is Not Present at {self.quantized_pth_location}")
```
Comment (shwu-nyunai) [inline quant/tensorrt/qat/main.py:123]: same as above. the exception is logged in the logger. logging and then raising exception will make the logger redundant. we can avoid unnecessary logs.

File: core/utils/modelutils.py [hunk 10]
```diff
@@ -119,20 +119,22 @@ def modify_head_classification(model, model_name, num_classes):
...
+def get_state_dict_or_model(loaded):
     """
-    Identifies if loaded model is a state dict or a nn.module
+    Identifies if loaded file is a state dict or a nn.module
     """
-    if isinstance(loaded, OrderedDict) or isinstance(loaded, dict):
+    if isinstance(loaded, OrderedDict, dict):
         return "STATE_DICT"
-    else:
+    elif isinstance(loaded, nn.Module): ## add imports
```
Comment (shwu-nyunai) [inline core/utils/modelutils.py:128]: the comment can be removed.

**Review by shwu-nyunai — CHANGES_REQUESTED** [review]

**Review by shwu-nyunai — APPROVED** [review]
lgtm. 👍🏻

**Pull request merged**